tests/unit/log_processing_test.py

```diff
@@ -270,6 +340,179 @@ def test_halt_with_matcher(self):
+
+    def test_cri_empty_lines(self):
+        timestamp_gen = timestamp_generator(10)
+        self.log_file = self._create_iterator(
+            {"path": self.__path, "parse_format": "cri"}
+        )
+        self.scan_for_new_bytes()
+
+        lines = [
+            next(timestamp_gen).iso + b" stdout F full_content_1\n",
+            next(timestamp_gen).iso + b" stdout P \n",
+            next(timestamp_gen).iso + b" stdout F full_content_2\n",
+            next(timestamp_gen).iso + b" stdout F \n"
+        ]
+        self.append_file(self.__path, *lines)
+
+        self.scan_for_new_bytes()
+
+        assert self.readline().line == b"full_content_1\n"
+        assert self.readline().line == b"full_content_2\n"
+        assert self.readline().line == b"\n"
+
+    def test_cri_split_messages_mixed_streams(self):
+        self.log_file = self._create_iterator(
+            {"path": self.__path, "parse_format": "cri"}
+        )
+        self.scan_for_new_bytes()
+
+        lines = [
+            b"2023-12-20T16:18:41.129027195+00:00 stdout F stdout: line 1\n",
+            b"2023-12-20T16:18:41.129027195+00:00 stdout P BEGIN PARTIAL_MESSAGE\n",
+            b"2023-12-20T16:18:41.129069776+00:00 stderr F stderr: line 1\n",
+            b"2023-12-20T16:18:41.129069776+00:00 stderr P BEGIN PARTIAL_MESSAGE\n",
+            b"2023-12-20T16:18:42.129069776+00:00 stderr P -MIDDLE OF PARTIAL_MESSAGE\n",
+            b"2023-12-20T16:18:43.186370550+00:00 stdout F -END PARTIAL_MESSAGE\n",
+            b"2023-12-20T16:18:43.187212946+00:00 stderr F -END PARTIAL_MESSAGE\n",
+            b"2023-12-20T16:18:43.194095421+00:00 stdout F stdout: line 2\n",
+        ]
+
+        self.append_file(self.__path, *lines)
+
+        self.scan_for_new_bytes()
+
+        assert_line(self.readline(), b"stdout: line 1\n", 1703089121129027195, "stdout")
+        assert_line(self.readline(), b"stderr: line 1\n", 1703089121129069776, "stderr")
+        assert_line(self.readline(), b"BEGIN PARTIAL_MESSAGE-END PARTIAL_MESSAGE\n", 1703089121129027195, "stdout")
+        assert_line(self.readline(), b"BEGIN PARTIAL_MESSAGE-MIDDLE OF PARTIAL_MESSAGE-END PARTIAL_MESSAGE\n", 1703089121129069776, "stderr")
+        assert_line(self.readline(), b"stdout: line 2\n", 1703089123194095421, "stdout")
+
+    def test_cri_split_messages_mixed_streams_small_page_size(self):
+        self.log_file = self._create_iterator(
+            {"path": self.__path, "parse_format": "cri"}
+        )
+        self.scan_for_new_bytes()
+
+        lines = [
+            b"2023-12-20T16:18:41.129027195+00:00 stdout F stdout: line 1\n",
+            b"2023-12-20T16:18:41.129027195+00:00 stdout P BEGIN PARTIAL_MESSAGE\n",
+            b"2023-12-20T16:18:41.129027195+00:00 stdout P -MIDDLE OF PARTIAL_MESSAGE\n",
+            b"2023-12-20T16:18:41.129069776+00:00 stderr F stderr: line 1\n",
+            b"2023-12-20T16:18:41.129069776+00:00 stderr P BEGIN PARTIAL_MESSAGE\n",
```

Comment:
In case partial message is split across multiple lines, we use timestamp of the first partial message, right?

Reply:
Yes.